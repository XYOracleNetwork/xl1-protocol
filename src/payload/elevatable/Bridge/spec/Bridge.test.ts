import { toAddress, toHex } from '@xylabs/hex'
import {
  describe, expect, it,
} from 'vitest'

import type { Chain } from '../../../../model.ts'
import { AttoXL1ConvertFactor } from '../../../../xl1/index.ts'
import type { BridgeIntentFields } from '../BridgeIntent.ts'
import type { BridgeObservationFields } from '../BridgeObservation.ts'

describe('Bridge', () => {
  const srcAmount = toHex(18n * AttoXL1ConvertFactor.xl1) // 18 XL1 in AttoXL1
  const destAmount = srcAmount // 1:1 for test

  const xl1ChainId: Chain = toHex('dd381fbb392c85160d8b0453e446757b12384046')
  const ethChainId = toHex('0x01')

  const xl1Sender = toAddress('1111111111111111111111111111111111111111')
  const ethReceiver = toAddress('0x2222222222222222222222222222222222222222')

  const bridgeableTokenContract = toHex('0x3333333333333333333333333333333333333333')

  const nonce = 'd5eeff33-d5bc-4aca-9ecb-3406c02a5dc4'

  const ethTxConfirmationHash = toHex('0x4444444444444444444444444444444444444444444444444444444444444444') // Some Eth tx hash

  describe('Bridge to ETH', () => {
    it('BridgeIntent', () => {
      const intent: BridgeIntentFields = {
        dest: ethChainId, // To Ethereum
        destAddress: ethReceiver,
        destAmount,
        destToken: bridgeableTokenContract,
        nonce,
        src: xl1ChainId, // From XL1
        srcAddress: xl1Sender, // From XL1 sender
        srcAmount,
        srcToken: xl1ChainId, // In XL1
      }
      expect(intent).toMatchSnapshot()
    })
    it('BridgeObservation', () => {
      const observation: BridgeObservationFields = {
        dest: ethChainId, // To Ethereum
        destAddress: ethReceiver,
        destAmount,
        destToken: bridgeableTokenContract,
        destConfirmation: ethTxConfirmationHash,
        src: xl1ChainId, // From XL1
        srcAddress: xl1Sender, // From XL1 sender
        srcAmount,
        srcToken: xl1ChainId, // In XL1
      }
      expect(observation).toMatchSnapshot()
    })
  })
})
