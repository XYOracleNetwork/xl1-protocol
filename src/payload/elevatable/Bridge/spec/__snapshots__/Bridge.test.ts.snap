// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Bridge > Bridge to ETH > BridgeIntent 1`] = `
{
  "dest": "01",
  "destAddress": "2222222222222222222222222222222222222222",
  "destAmount": "f9ccd8a1c5080000",
  "destToken": "3333333333333333333333333333333333333333",
  "nonce": "d5eeff33-d5bc-4aca-9ecb-3406c02a5dc4",
  "src": "dd381fbb392c85160d8b0453e446757b12384046",
  "srcAddress": "1111111111111111111111111111111111111111",
  "srcAmount": "f9ccd8a1c5080000",
  "srcToken": "dd381fbb392c85160d8b0453e446757b12384046",
}
`;

exports[`Bridge > Bridge to ETH > BridgeObservation 1`] = `
{
  "dest": "01",
  "destAddress": "2222222222222222222222222222222222222222",
  "destAmount": "f9ccd8a1c5080000",
  "destConfirmation": "4444444444444444444444444444444444444444444444444444444444444444",
  "destToken": "3333333333333333333333333333333333333333",
  "src": "dd381fbb392c85160d8b0453e446757b12384046",
  "srcAddress": "1111111111111111111111111111111111111111",
  "srcAmount": "f9ccd8a1c5080000",
  "srcToken": "dd381fbb392c85160d8b0453e446757b12384046",
}
`;
